### ISSUE TITLE
hide unimplemented syntax from diagrams

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/docs#6659]
There are various cases where the syntax supports a construct but it immediately errors out with "not implemented". We should make sure these are hidden from diagrams. Perhaps it's as easy as adding `/* SKIP DOC */` where necessary?

An example is the `ON CONFLICT WHERE` construct, which is not supported:

![image](https://user-images.githubusercontent.com/16544120/75073828-4d94a100-54c8-11ea-9baa-7a2d90e2e5c8.png)

**Comment by jseldess** [comment]
@RaduBerinde, how do we track down these cases? I agree that it's probably best to hide these from sql diagrams via `/* SKIP DOC */`, but if that's not possible, we can adjust the way we generate the diagrams. In the meantime, I'll remove that specific where clause.

**Comment by RaduBerinde** [comment]
I'd grep for `unimplemented` in `sql.y` to find other cases.

**Comment by jseldess** [comment]
Opened a PR to manually hide the where clause in the on conflict diagram, but that's not the best approach. Instead, we should be able to use `/* SKIP DOC */` to tell cockroach not to include this syntax in generated bnfs in the first place. In principle, this should be possible, or happen by default, for all unimplemented syntax. It's not clear how to go about that in this and probably lots of other cases, however. Will discuss with @knz once he's back.

**Comment by knz** [comment]
> Perhaps it's as easy as adding `/* SKIP DOC */` where necessary?

Yes that seems reasonable to me. What I think happened is that the engineer(s) who added the extra clause with the "unimplemented" action (for telemetry purposes) simply forgot to add the "SKIP DOC" marker. I know I have forgotten to do this in the past, and I've seen others forget too.

**Comment by jseldess** [comment]
This depends on a change in cockroach: https://github.com/cockroachdb/cockroach/issues/45464. If/when that happens, we'll need to understand which bnfs are effected and regenerate them for our docs.

**Comment by jseldess** [comment]
@ianjevans, since you're leading the effort to move from SQL statement diagrams to text-based syntax definitions, I'll assign this to you. Ideally, in the text-based approach, we should avoid featuring expressions that are supported but "unimplemented".

**Comment by knz** [comment]
So there are two types of situations:
- cases where we parse the syntax and report an error immediately during parsing (with a link to a github issue). These can be recognized with e.g. `return unimplemented(...)` in `sql.y`. I can confirm that many of those do not have "SKIP DOC" associated with them, and we could do that.  We could also enhance the diagram generator to automatically skip rules that contain only "return unimplemented". I'll give a shot at that.

- However, meanwhile, there are many cases where we parse the syntax and the error is generated much later, during query planning. The example by @RaduBerinde's at the top, `ON CONFLICT(...) WHERE`, alas, is one of these situations. For these, we cannot easily match the syntax to the error; the relationship is very distant, as it needs the investigator to map the error in the query planning code to the abstract syntax tree back to the grammar. This investigation would be more involved and would need help from someone from the SQL queries team.

**Comment by knz** [comment]
> I can confirm that many of those do not have "SKIP DOC" associated with them, and we could do that. We could also enhance the diagram generator to automatically skip rules that contain only "return unimplemented". I'll give a shot at that.

This logic is already there. The diagram generator already skips those.

This means the only cases remaining are the "hard" cases for which looking at the grammar is not sufficient. I don't think we can offload this to someone on the docs team. This needs an investigation by SQL experience/queries.

**Comment by knz** [comment]
yes
